core/src/main/java/com/datastax/oss/driver/api/core/cql/PagingState.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.cql;
+
+import com.datastax.oss.driver.api.core.session.Session;
+import com.datastax.oss.driver.internal.core.cql.DefaultPagingState;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.ByteBuffer;
+
+/**
+ * A safe wrapper around the paging state of a query.
+ *
+ * <p>This class performs additional checks to fail fast if the paging state is not reused on the
+ * same query, and it provides utility methods for conversion to/from strings and byte arrays.
+ */
+public interface PagingState {
+
+  /** Parses an instance from a string previously generated with {@link #toString()}. */
+  @NonNull
+  static PagingState fromString(@NonNull String string) {
+    return DefaultPagingState.fromString(string);
+  }
+
+  /** Parses an instance from a byte array previously generated with {@link #toBytes()}. */
+  @NonNull
+  static PagingState fromBytes(byte[] bytes) {
+    return DefaultPagingState.fromBytes(bytes);
+  }
+
+  /** Returns a representation of this object as a byte array. */
+  byte[] toBytes();
+
+  /**
+   * Checks if this paging state can be safely reused for the given statement. Specifically, the
+   * query string and any bound values must match.
+   *
+   * <p>Note that, if {@code statement} is a {@link SimpleStatement} with bound values, those values
+   * must be encoded in order to perform the check. This method uses the default codec registry and
+   * default protocol version. This might fail if you use custom codecs; in that case, use {@link
+   * #matches(Statement, Session)} instead.
```

Comment:
That whole thing with simple statements is annoying, it adds a bunch of complexity for a minuscule edge case. Maybe we should have attached simple statements to the session (or gotten rid of them altogether), but that would have been yet another difference compared to driver 3.